Let the compiler generate more comparison operators in WebCore

cdumez opened the pull request:
#### c14146dd49a0b019dccb408d52339caa77594558
<pre>
Let the compiler generate more comparison operators in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=261095">https://bugs.webkit.org/show_bug.cgi?id=261095</a>

Reviewed by Darin Adler.

Let the compiler generate more comparison operators in WebCore now
that we support c++20.

* Source/WebCore/rendering/CSSValueKey.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/ClipRect.h:
(WebCore::ClipRect::operator== const): Deleted.
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegion::operator== const): Deleted.
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/GapRects.h:
(WebCore::GapRects::operator== const): Deleted.
* Source/WebCore/rendering/HighlightData.h:
(WebCore::RenderRange::operator== const): Deleted.
* Source/WebCore/rendering/LayerAncestorClippingStack.h:
(WebCore::CompositedClipData::operator== const): Deleted.
* Source/WebCore/rendering/Pagination.h:
(WebCore::Pagination::operator== const): Deleted.
* Source/WebCore/rendering/shapes/ShapeInterval.h:
(WebCore::ShapeInterval::operator== const): Deleted.
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::SVGPathTransformedByteStream::operator== const): Deleted.
* Source/WebCore/rendering/style/BorderData.h:
(WebCore::BorderData::operator== const): Deleted.
* Source/WebCore/rendering/style/BorderValue.h:
(WebCore::BorderValue::operator== const): Deleted.
* Source/WebCore/rendering/style/CounterContent.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/CounterDirectives.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/CursorData.h:
(WebCore::CursorData::operator== const): Deleted.
* Source/WebCore/rendering/style/FillLayer.h:
(WebCore::FillRepeatXY::operator== const): Deleted.
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/GapLength.h:
(WebCore::GapLength::operator== const): Deleted.
* Source/WebCore/rendering/style/GridArea.h:
(WebCore::GridSpan::operator== const): Deleted.
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/GridLength.h:
(WebCore::GridLength::operator== const): Deleted.
* Source/WebCore/rendering/style/GridPosition.cpp:
(WebCore::GridPosition::operator== const): Deleted.
* Source/WebCore/rendering/style/GridPosition.h:
* Source/WebCore/rendering/style/LineClampValue.h:
(WebCore::LineClampValue::operator== const): Deleted.
* Source/WebCore/rendering/style/ListStyleType.h:
(WebCore::ListStyleType::operator== const): Deleted.
* Source/WebCore/rendering/style/NinePieceImage.h:
(WebCore::NinePieceImage::operator== const): Deleted.
* Source/WebCore/rendering/style/OffsetRotation.cpp:
(WebCore::OffsetRotation::operator== const): Deleted.
* Source/WebCore/rendering/style/OffsetRotation.h:
* Source/WebCore/rendering/style/OutlineValue.h:
(WebCore::OutlineValue::operator== const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::InheritedFlags::operator== const): Deleted.
(WebCore::RenderStyle::NonInheritedFlags::operator== const): Deleted.
* Source/WebCore/rendering/style/SVGRenderStyle.h:
(WebCore::SVGRenderStyle::InheritedFlags::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleColorScheme.h:
(WebCore::StyleColorScheme::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleContentAlignmentData.h:
(WebCore::StyleContentAlignmentData::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::StyleGridData::operator== const):
(WebCore::operator==): Deleted.
(WebCore::GridTrackEntrySubgrid::operator== const): Deleted.
(WebCore::GridTrackEntryMasonry::operator== const): Deleted.
(WebCore::GridTrackEntryRepeat::operator== const): Deleted.
(WebCore::GridTrackEntryAutoRepeat::operator== const): Deleted.
(WebCore::MasonryAutoFlow::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleScrollSnapPoints.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
(WebCore::StyleSelfAlignmentData::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleTextBoxEdge.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/TextSizeAdjustment.h:
(WebCore::TextSizeAdjustment::operator== const): Deleted.
(WebCore::AutosizeStatus::operator== const): Deleted.
* Source/WebCore/rendering/style/TextUnderlineOffset.h:
(WebCore::TextUnderlineOffset::operator== const): Deleted.
* Source/WebCore/rendering/style/WillChangeData.h:
(WebCore::WillChangeData::AnimatableFeature::operator== const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.h:
(WebCore::GradientData::Inputs::operator== const): Deleted.
* Source/WebCore/style/MatchResult.h:
(WebCore::Style::MatchResult::isEmpty const):
* Source/WebCore/svg/SVGLengthValue.h:
(WebCore::operator==): Deleted.
* Source/WebCore/svg/SVGPathByteStream.h:
(WebCore::SVGPathByteStream::operator== const): Deleted.
* Source/WebCore/workers/service/ServiceWorkerJobDataIdentifier.h:
(WebCore::operator==): Deleted.
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.cpp:
(WebCore::ServiceWorkerRegistrationKey::operator== const): Deleted.
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/shared/SharedWorkerKey.h:
(WebCore::operator==): Deleted.

Canonical link: <a href="https://commits.webkit.org/267612@main">https://commits.webkit.org/267612@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a58a335d81849c9198dc35d7c2b9872e294dc40a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17182 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17508 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18003 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18969 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16080 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17374 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20782 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17650 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18275 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17729 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14920 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19786 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22297 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15968 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20115 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16369 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13887 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15523 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4098 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19890 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16200 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/StyleGridData.h:
```diff
@@ -40,9 +40,10 @@ namespace WebCore {
 };
 
-inline bool operator==(const NamedGridLinesMap& a, const NamedGridLinesMap& b) { return a.map == b.map; }
 inline bool operator!=(const NamedGridLinesMap& a, const NamedGridLinesMap& b) { return a.map != b.map; }
```
We should be able to remove this without breaking anything since C++20 defines `!=` based on `==` automatically.

darinadler commented on Source/WebCore/rendering/style/StyleGridData.h:
```diff
@@ -115,7 +101,7 @@ class StyleGridData : public RefCounted<StyleGridData> {
     bool operator==(const StyleGridData& o) const
     {
-        return m_columns.list == o.m_columns.list && m_rows.list == o.m_rows.list && implicitNamedGridColumnLines.map == o.implicitNamedGridColumnLines.map && implicitNamedGridRowLines.map == o.implicitNamedGridRowLines.map && gridAutoFlow == o.gridAutoFlow && gridAutoRows == o.gridAutoRows && gridAutoColumns == o.gridAutoColumns && namedGridArea.map == o.namedGridArea.map && namedGridAreaRowCoun ...
+        return m_columns == o.m_columns && m_rows == o.m_rows && implicitNamedGridColumnLines == o.implicitNamedGridColumnLines && implicitNamedGridRowLines == o.implicitNamedGridRowLines && gridAutoFlow == o.gridAutoFlow && gridAutoRows == o.gridAutoRows && gridAutoColumns == o.gridAutoColumns && namedGridArea == o.namedGridArea && namedGridAreaRowCount == o.namedGridAreaRowCount && namedGridAre ...
```
The changes here are fine, but this function worries me. It’s quite surprising what is checked and what is not and not at all clear that it’s all correct. We may want to refactor this class to separate computed data from the data it’s computed from.

darinadler commented on Source/WebCore/rendering/style/TextSizeAdjustment.h:
```diff
@@ -42,7 +42,7 @@ class TextSizeAdjustment {
     constexpr bool isPercentage() const { return m_value >= 0; }
 
-    constexpr bool operator==(const TextSizeAdjustment& anAdjustment) const { return m_value == anAdjustment.m_value; }
+    friend constexpr bool operator==(const TextSizeAdjustment&, const TextSizeAdjustment&) = default;
```
Since this is just a `float`, clearly it can just be `TextSizeAdjustment` without the `const&`, but I am unclear on the best rule we should follow about when to do that. I haven’t commented every time I see the possibility of omitting `const&`.

darinadler commented on Source/WebCore/rendering/style/TextSizeAdjustment.h:
```diff
@@ -64,7 +64,7 @@ class AutosizeStatus {
     constexpr bool contains(Fields) const;
 
-    constexpr bool operator==(const AutosizeStatus& other) const { return fields() == other.fields(); }
+    friend constexpr bool operator==(const AutosizeStatus&, const AutosizeStatus&) = default;
```
Since this is just an `OptionSet`, clearly it can just be `AutosizeStatus` without the `const&`, but I am unclear on the best rule we should follow about when to do that.

darinadler commented on Source/WebCore/rendering/style/TextUnderlineOffset.h:
```diff
@@ -69,10 +69,7 @@ class TextUnderlineOffset {
-    {
-        return m_length == other.m_length;
-    }
+    friend constexpr bool operator==(const TextUnderlineOffset&, const TextUnderlineOffset&) = default;
```
Just a `std::optional<float>`, no `const&` needed.

darinadler commented on Source/WebCore/svg/SVGLengthValue.h:
```diff
@@ -95,17 +95,14 @@ class SVGLengthValue {
 
     ExceptionOr<void> convertToSpecifiedUnits(const SVGLengthContext&, SVGLengthType);
 
+    friend bool operator==(const SVGLengthValue&, const SVGLengthValue&) = default;
```
Just a `float` and a couple scalars. No `const&` needed.

cdumez committed:
Let the compiler generate more comparison operators in WebCore
https://bugs.webkit.org/show_bug.cgi?id=261095

Reviewed by Darin Adler.

Let the compiler generate more comparison operators in WebCore now
that we support c++20.

* Source/WebCore/rendering/CSSValueKey.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/ClipRect.h:
(WebCore::ClipRect::operator== const): Deleted.
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegion::operator== const): Deleted.
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/GapRects.h:
(WebCore::GapRects::operator== const): Deleted.
* Source/WebCore/rendering/HighlightData.h:
(WebCore::RenderRange::operator== const): Deleted.
* Source/WebCore/rendering/LayerAncestorClippingStack.h:
(WebCore::CompositedClipData::operator== const): Deleted.
* Source/WebCore/rendering/Pagination.h:
(WebCore::Pagination::operator== const): Deleted.
* Source/WebCore/rendering/shapes/ShapeInterval.h:
(WebCore::ShapeInterval::operator== const): Deleted.
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::SVGPathTransformedByteStream::operator== const): Deleted.
* Source/WebCore/rendering/style/BorderData.h:
(WebCore::BorderData::operator== const): Deleted.
* Source/WebCore/rendering/style/BorderValue.h:
(WebCore::BorderValue::operator== const): Deleted.
* Source/WebCore/rendering/style/CounterContent.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/CounterDirectives.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/CursorData.h:
(WebCore::CursorData::operator== const): Deleted.
* Source/WebCore/rendering/style/FillLayer.h:
(WebCore::FillRepeatXY::operator== const): Deleted.
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/GapLength.h:
(WebCore::GapLength::operator== const): Deleted.
* Source/WebCore/rendering/style/GridArea.h:
(WebCore::GridSpan::operator== const): Deleted.
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/GridLength.h:
(WebCore::GridLength::operator== const): Deleted.
* Source/WebCore/rendering/style/GridPosition.cpp:
(WebCore::GridPosition::operator== const): Deleted.
* Source/WebCore/rendering/style/GridPosition.h:
* Source/WebCore/rendering/style/LineClampValue.h:
(WebCore::LineClampValue::operator== const): Deleted.
* Source/WebCore/rendering/style/ListStyleType.h:
(WebCore::ListStyleType::operator== const): Deleted.
* Source/WebCore/rendering/style/NinePieceImage.h:
(WebCore::NinePieceImage::operator== const): Deleted.
* Source/WebCore/rendering/style/OffsetRotation.cpp:
(WebCore::OffsetRotation::operator== const): Deleted.
* Source/WebCore/rendering/style/OffsetRotation.h:
* Source/WebCore/rendering/style/OutlineValue.h:
(WebCore::OutlineValue::operator== const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::InheritedFlags::operator== const): Deleted.
(WebCore::RenderStyle::NonInheritedFlags::operator== const): Deleted.
* Source/WebCore/rendering/style/SVGRenderStyle.h:
(WebCore::SVGRenderStyle::InheritedFlags::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleColorScheme.h:
(WebCore::StyleColorScheme::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleContentAlignmentData.h:
(WebCore::StyleContentAlignmentData::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::StyleGridData::operator== const):
(WebCore::operator==): Deleted.
(WebCore::GridTrackEntrySubgrid::operator== const): Deleted.
(WebCore::GridTrackEntryMasonry::operator== const): Deleted.
(WebCore::GridTrackEntryRepeat::operator== const): Deleted.
(WebCore::GridTrackEntryAutoRepeat::operator== const): Deleted.
(WebCore::MasonryAutoFlow::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleScrollSnapPoints.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
(WebCore::StyleSelfAlignmentData::operator== const): Deleted.
* Source/WebCore/rendering/style/StyleTextBoxEdge.h:
(WebCore::operator==): Deleted.
* Source/WebCore/rendering/style/TextSizeAdjustment.h:
(WebCore::TextSizeAdjustment::operator== const): Deleted.
(WebCore::AutosizeStatus::operator== const): Deleted.
* Source/WebCore/rendering/style/TextUnderlineOffset.h:
(WebCore::TextUnderlineOffset::operator== const): Deleted.
* Source/WebCore/rendering/style/WillChangeData.h:
(WebCore::WillChangeData::AnimatableFeature::operator== const): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceGradient.h:
(WebCore::GradientData::Inputs::operator== const): Deleted.
* Source/WebCore/style/MatchResult.h:
(WebCore::Style::MatchResult::isEmpty const):
* Source/WebCore/svg/SVGLengthValue.h:
(WebCore::operator==): Deleted.
* Source/WebCore/svg/SVGPathByteStream.h:
(WebCore::SVGPathByteStream::operator== const): Deleted.
* Source/WebCore/workers/service/ServiceWorkerJobDataIdentifier.h:
(WebCore::operator==): Deleted.
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.cpp:
(WebCore::ServiceWorkerRegistrationKey::operator== const): Deleted.
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/shared/SharedWorkerKey.h:
(WebCore::operator==): Deleted.

Canonical link: https://commits.webkit.org/267612@main
